### PR TITLE
feat: set proper grouping hash for all errors

### PR DESCRIPTION
**Pull request by saikumarrs**
## PR Description

Please include a summary of the change along with the relevant motivation and context.

## Linear task (optional)

https://linear.app/rudderstack/issue/SDK-3304/fix-bugsnag-grouping-for-error-messages-with-ids

## Cross Browser Tests

Please confirm you have tested for the following browsers:

- [ ] Chrome
- [ ] Firefox
- [ ] IE11

## Sanity Suite

- [ ] All sanity suite test cases pass locally

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

## Summary by CodeRabbit

- **Refactor**
  - Standardized error handling across the system to use structured objects with detailed context, custom messages, and optional grouping information.
  - Updated error messages for improved clarity, including consistent terminology changes from "destination" to "integration."
  - Enhanced error reporting for script loading, plugin management, consent managers, and event forwarding with richer context and grouping support.
  - Simplified external script loader initialization by removing redundant error handler dependencies.
  - Adjusted method signatures and interfaces to support the new error handling structure and improved error grouping.
  - Improved error message formatting and added new error message variants for integration readiness and transformation failures.
  - Improved error grouping logic with a new utility for normalized grouping hashes.
- **Tests**
  - Updated all relevant tests to match the new error handling interface and improved error message expectations.
  - Added new tests covering exception and timeout scenarios during script loading.
  - Added tests for destination SDK initialization and integration readiness utilities.
  - Added tests for event forwarding with error handling in native destination queue utilities.
  - Added tests for error grouping hash utility function.
  - Added comprehensive new test suites for BeaconQueue and NativeDestinationQueue plugins covering initialization, event processing, error handling, and queue management.
  - Expanded device mode transformation utilities tests for success, failure, and edge cases.
  - Enhanced CapabilitiesManager tests with extensive mocks and coverage for capability detection, polyfill handling, event listeners, and lifecycle management.
  - Extended DeviceModeDestinations plugin tests with mocks, error handling, configuration overrides, and loading scenarios.
  - Improved Store class tests with new scenarios for storage errors, circular references, and fallback handling.
  - Added Analytics lifecycle error handling test.
  - Added PluginsManager registration and unregistration error handling tests.
  - Added UserSessionManager cookie setting error handling test.
  - Enhanced ConfigManager and ErrorHandler tests for updated error handling signatures.
- **Chores**
  - Increased allowed bundle size limits for several build targets.
- **New Features**
  - Added a new slow response handler in test fixtures to simulate network timeout scenarios.

### CODE REVIEW

File: packages/analytics-js-plugins/__tests__/deviceModeTransformation/utilities.test.ts [hunk 1]
```diff
@@ -1,22 +1,461 @@
...
+        sampleEvent,
+        mockErrorHandler,
+        mockLogger,
+      );
+
+      expect(mockErrorHandler.onError).toHaveBeenCalledWith({
+        error: expect.any(Error),
+        context: 'DeviceModeTransformationPlugin',
+        customMessage: expect.stringContaining('GA4'),
+      });
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-plugins/__tests__/deviceModeTransformation/utilities.test.ts:353]: It was intentionally left out.

File: packages/analytics-js/src/services/ErrorHandler/utils.ts [hunk 2]
```diff
@@ -206,6 +209,32 @@
...
+    } else {
+      normalizedGroupingHash = defaultGroupingHash;
+    }
+  } else if (groupingHash === '') {
+    normalizedGroupingHash = defaultGroupingHash;
+  } else {
+    normalizedGroupingHash = groupingHash;
+  }
+  return normalizedGroupingHash;
+};
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js/src/services/ErrorHandler/utils.ts]: Addressed.

File: packages/analytics-js-plugins/src/deviceModeTransformation/logMessages.ts [hunk 3]
```diff
@@ -16,7 +16,9 @@
 ): string =>
   `${context}${LOG_CONTEXT_SEPARATOR}[Destination: ${displayName}].Transformation request failed with status: ${status}. Retries exhausted. ${action}.`;
 
-const DMT_EXCEPTION = (displayName: string): string => `[Destination:${displayName}].`;
+const DMT_EXCEPTION = (displayName: string): string =>
+  `Transformation failed for destination "${displayName}"`;
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-plugins/src/deviceModeTransformation/logMessages.ts:20]: Addressed.

File: packages/analytics-js-plugins/src/beaconQueue/logMessages.ts [hunk 4]
```diff
@@ -9,11 +9,11 @@
 const BEACON_QUEUE_BLOB_CONVERSION_FAILURE_ERROR = (context: string): string =>
   `${context}${LOG_CONTEXT_SEPARATOR}Failed to convert events batch object to Blob.`;
 
-const BEACON_QUEUE_SEND_ERROR = (context: string): string =>
-  `${context}${LOG_CONTEXT_SEPARATOR}Failed to send events batch data to the browser's beacon queue. The events will be dropped.`;
+const BEACON_QUEUE_SEND_ERROR = (context: string, url: string): string =>
+  `${context}${LOG_CONTEXT_SEPARATOR}${BEACON_QUEUE_DELIVERY_ERROR(url)}`;
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-plugins/src/beaconQueue/logMessages.ts:13]: Addressed.

**Comment by saikumarrs** [comment]
BugBot run

**Comment by saikumarrs** [comment]
BugBot run

**Comment by saikumarrs** [comment]
BugBot run

File: packages/analytics-js-plugins/src/deviceModeDestinations/index.ts [hunk 5]
```diff
@@ -133,12 +136,8 @@ const DeviceModeDestinations = (): ExtensionPlugin => ({
               }),
             timeout: SCRIPT_LOAD_TIMEOUT_MS,
           });
-        } else if (sdkTypeName) {
-          initializeDestination(dest, state, destSDKIdentifier, sdkTypeName, errorHandler, logger);
         } else {
-          logger?.error(
-            DESTINATION_SDK_LOAD_ERROR(DEVICE_MODE_DESTINATIONS_PLUGIN, dest.displayName),
-          );
```
Comment (MoumitaM) [inline packages/analytics-js-plugins/src/deviceModeDestinations/index.ts:138]: What is the motivation behind removing this else if check?

File: packages/analytics-js/src/components/pluginsManager/PluginsManager.ts [hunk 6]
```diff
@@ -273,7 +273,7 @@ class PluginsManager implements IPluginsManager {
               ...state.plugins.failedPlugins.value,
               remotePluginKey,
             ];
-            this.onError(err, remotePluginKey);
+            this.onError(err, `Failed to load plugin "${remotePluginKey}"`, err);
```
Comment (MoumitaM) [inline packages/analytics-js/src/components/pluginsManager/PluginsManager.ts:276]: groupingHash is expected as string but we are sending err.

File: packages/analytics-js/src/services/ErrorHandler/utils.ts [hunk 7]
```diff
@@ -206,6 +209,44 @@ const getErrorDeliveryPayload = (payload: ErrorEventPayload, state: ApplicationS
...
+  state: ApplicationState,
+  logger: ILogger,
+) => {
+  let normalizedGroupingHash: string | undefined;
+  if (state.context.app.value.installType !== 'cdn') {
+    return normalizedGroupingHash;
+  }
+
+  if (!isString(curErrGroupingHash)) {
+    const normalizedErrorInstance = normalizeError(curErrGroupingHash, logger);
```
Comment (MoumitaM) [inline packages/analytics-js/src/services/ErrorHandler/utils.ts:240]: In case of undefined `curErrGroupingHash` `normalizeError` will log some warning which is not correct.

**Review by MoumitaM — CHANGES_REQUESTED** [review]

File: packages/analytics-js-plugins/src/deviceModeDestinations/index.ts (hunk 5 above)
Comment (saikumarrs) [reply packages/analytics-js-plugins/src/deviceModeDestinations/index.ts:138]: The else block was redundant as the destination support check is already done previously.

File: packages/analytics-js/src/components/pluginsManager/PluginsManager.ts (hunk 6 above)
Comment (saikumarrs) [reply packages/analytics-js/src/components/pluginsManager/PluginsManager.ts:276]: For special instances, we are sending the error object itself to properly use the exact error message for grouping.
The exact error message deduction is done within the ErrorHandler.

File: packages/analytics-js/src/services/ErrorHandler/utils.ts (hunk 7 above)
Comment (saikumarrs) [reply packages/analytics-js/src/services/ErrorHandler/utils.ts:240]: If the error instance is not a valid error, it would validated with a warning without even reaching this step.

File: packages/analytics-js/src/components/capabilitiesManager/CapabilitiesManager.ts [hunk 8]
```diff
@@ -200,8 +200,12 @@
...
    */
-  onError(error: unknown): void {
-    this.errorHandler.onError(error, CAPABILITIES_MANAGER);
+  onError(error: unknown, groupingHash?: string): void {
+    this.errorHandler.onError({
+      error,
+      context: CAPABILITIES_MANAGER,
+      groupingHash,
+    });
   }
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js/src/components/capabilitiesManager/CapabilitiesManager.ts:209]: Addressed.

File: packages/analytics-js/__tests__/components/capabilitiesManager/CapabilitiesManager.test.ts [hunk 9]
```diff
@@ -31,12 +65,155 @@ jest.mock('../../../src/components/capabilitiesManager/polyfill', () => {
...
+      innerWidth: 1920,
+      innerHeight: 1080,
+    });
+    mockIsStorageAvailable.mockReturnValue(true);
+    mockGetUserAgent.mockReturnValue('test-user-agent');
+    mockGetLanguage.mockReturnValue('en-US');
+    mockGetTimezone.mockReturnValue('America/New_York');
+    // @ts-expect-error needed for the test
+    POLYFILL_URL = 'https://somevalid.polyfill.url';
+  });
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js/__tests__/components/capabilitiesManager/CapabilitiesManager.test.ts:129]: Addressed.

File: packages/analytics-js/__tests__/components/pluginsManager/PluginsManager.test.ts [hunk 10]
```diff
@@ -12,6 +12,331 @@ describe('PluginsManager', () => {
...
+
+      // Verify the failing plugin was added to failed plugins
+      expect(state.plugins.failedPlugins.value).toEqual([...initialFailedPlugins, 'FailingPlugin']);
+
+      // Verify error handler was called with correct parameters
+      expect(errorHandlerSpy).toHaveBeenCalledWith({
+        error: expect.any(Error),
+        context: 'PluginsManager',
+        customMessage: 'Failed to register plugin "FailingPlugin"',
+      });
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js/__tests__/components/pluginsManager/PluginsManager.test.ts]: Addressed.

**Review by MoumitaM — APPROVED** [review]
